### PR TITLE
Pioneer DDJ-400

### PR DESCRIPTION
**Pull request by nschloe**
After #2143 and #2312, this is another take on the Pioneer DDJ-400.

This PR implements most of the basic functionality of the controller:

  * Mixer Section (Faders, EQ, Filter, Gain, Cue)
  * Browsing and loading
  * Jogwheels, Scratching, Bending
  * Beat Sync
  * Beatjump mode
  * Hot cue Mode
  * Beat Loop Mode

I've successfully used this on two occasions and only found minor glitches. For example, when a hot cue point is set it lights up (this is good), but when shift is pressed, the lights go out. The functionality is still there though.

There are still a bunch of advanced controls missing, for example the entire effects section. This can be added in second PR though.

The visual XML composer was of great help creating this PR as was @WarkerAnhaltRanger first pass.

I'd suggest we keep it simple here and try to add more functionality in follow-up PRs. I'd be happy about some feedback of course.

### CODE REVIEW

**Comment by Be-ing** [comment]
> when a hot cue point is set it lights up (this is good), but when shift is pressed, the lights go out.

You may need to send a secondary MIDI message for the shifted state to prevent this. It is probably the same MIDI values as the controller sends when that button is pressed with shift.

**Comment by Be-ing** [comment]
Thanks for continuing to work on this. Can you update the wiki page with whatever changes you have made?

> There are still a bunch of advanced controls missing, for example the entire effects section. This can be added in second PR though.

The mapping should be complete before merging. Having a mapping included in Mixxx and listing it as supported on the wiki should be an indication to (potential) users that it just works. Otherwise we set users up for disappointment.

**Comment by nschloe** [comment]
> The mapping should be complete before merging.

For real? Then we can immediately cancel this. I won't be able to to get a complete mapping of the DDJ-400 up, this is just _so_ much work for me. I think, however, if we communicate clearly that not everything is working, it should be fine. There are multiple options for this:
  * not listing the DDJ-400 as supported
  * listing it at partially or basically supported
  * ...

**Comment by Be-ing** [comment]
What is still left to do? Is it only the effects?

**Comment by Holzhaus** [comment]
> What is still left to do? Is it only the effects?

https://github.com/mixxxdj/mixxx/pull/2385/files#diff-6a3e4882503838847f2b99ee50f272e0R19-R33

**Comment by nschloe** [comment]
> What is still left to do? Is it only the effects?

No, it's more than that. For example, the DDJ-400 has a fancy-schmancy shift button that enables all kind of functionality: "keyboard mode", key shift, etc. I consider this _advanced_ and I probably won't ever need any of that, so I didn't bother with it right now. Also, the first PR #2143 got turned down because it took so long for it to get reviewed that the submitted had lost interest. I wanted to avoid this here.

But, like I said, I'd be happy to develop this further in a separate PR.

**Comment by Be-ing** [comment]
I think we are talking about different ideas of "complete". I don't care about reimplementing everything from #2143. I'd also be okay with leaving the alternate pad modes (keyboard, pad FX (whatever that does), key shift) unmapped. I do care that everything on the controller does something useful that makes sense with Mixxx, which could be fairly different from how it works with Rekordbox. So how about mapping the samplers and effects, fixing the known bugs, and calling that good enough?

**Comment by Be-ing** [comment]
Here is a proposal for how to map the unusual effects section on this controller:
  * Beat `<` / `>` buttons: focus previous/next effect
  * FX select button: unfocus; make no effect focused
  * 1/2/Master switch: assign effect unit 1 to deck 1/2/master as labeled
  * Level/Depth knob: metaknob of focused effect; if no effect is focused, controls the mix knob for effect unit 1
  * On/Off button: enable/disable focused effect

**Comment by nschloe** [comment]
Okay, I'll keep working on it here then.

> You may need to send a secondary MIDI message for the shifted state to prevent this.

I don't understand what you mean here.

**Comment by Be-ing** [comment]
I see that the hotcue buttons send MIDI messages with a 1st byte (status) as 0x97 when pressed but it is 0x98 when shift is pressed. You only send outgoing MIDI for 0x97, but not for 0x98. You may need to send outgoing MIDI for 0x98 to prevent the LEDs from flickering when shift is pressed.

**Comment by nschloe** [comment]
Another thing I don't understand about development: mixxx always appears to create a file `DDJ-400_MIDI_1.midi.xml` in `~/.mixxx/controllers/`. 
```
ls -l ~/.mixxx/controllers/
```
```
-rw-r--r-- 1 nschloe nschloe  51K Dec  8 22:19 DDJ-400_MIDI_1.midi.xml
-rw-r--r-- 1 nschloe nschloe 527K Nov 26 17:20 lodash.mixxx.js
-rw-r--r-- 1 nschloe nschloe  48K Nov 26 16:40 midi-components-0.0.js
lrwxrwxrwx 1 nschloe nschloe   77 Nov 26 16:32 Pioneer-DDJ-400.midi.xml -> /home/nschloe/software/mixxx/nschloe/res/controllers/Pioneer-DDJ-400.midi.xml
lrwxrwxrwx 1 nschloe nschloe   78 Nov 26 16:32 Pioneer-DDJ-400-script.js -> /home/nschloe/software/mixxx/nschloe/res/controllers/Pioneer-DDJ-400-script.js
```
It seems to be a copy of the actual `Pioneer-DDJ-400.midi.xml`. This has caught be off-guard a couple of times, editing `Pioneer-DDJ-400.midi.xml` and wondering why nothing changes. Is this a bug?

**Comment by Be-ing** [comment]
It's not exactly a bug; it's more of a confusing workflow that could be improved. The "DDJ-400_MIDI_1.midi.xml" is automatically copied by Mixxx and the filename comes from the MIDI device name. That is the file Mixxx automatically loads on startup by seeing that there is a MIDI device which has the same name. Maybe we can rethink this together with the coming changes for controllers in 2.4.

**Comment by nschloe** [comment]
> I see that the hotcue buttons send MIDI messages with a 1st byte (status) as 0x97 when pressed but it is 0x98 when shift is pressed. You only send outgoing MIDI for 0x97, but not for 0x98. You may need to send outgoing MIDI for 0x98 to prevent the LEDs from flickering when shift is pressed.

The "flickering" happens _before_ 0x98 is sent. You press a hotcue, 0x97 is sent, mixxx sends a message to turn on the light, all fine. Then when shift is pressed, without ever touching a hotcue button, the lights go out.

**Comment by nschloe** [comment]
Another question: Is there any way of filtering the the controllerDebug midi codes? Now that blinking Cue is working, it's hard to find the messages I'm actually looking for in the sea of cue-on/cue-off messages.

**Comment by Be-ing** [comment]
Right, what it seems like is happening is that the controller firmware maintains separate states for the LEDs depending on whether shift is pressed. So if you want to avoid flickering when shift is pressed, both the controller's shifted and unshifted states need to be set whenever hotcue_X_enabled changes.

**Comment by nschloe** [comment]
> It's not exactly a bug; it's more of a confusing workflow that could be improved. The "DDJ-400_MIDI_1.midi.xml" is automatically copied by Mixxx and the filename comes from the MIDI device name. That is the file Mixxx automatically loads on startup by seeing that there is a MIDI device which has the same name. Maybe we can rethink this together with the coming changes for controllers in 2.4.

I'll file a bug for it so we don't lose track.

**Comment by nschloe** [comment]
> Right, what it seems like is happening is that the controller firmware maintains separate states for the LEDs depending on whether shift is pressed.

Indeed.

> So if you want to avoid flickering when shift is pressed, both the controller's shifted and unshifted states need to be set whenever hotcue_X_enabled changes.

I don't understand "setting the state". I press the shift button and the controller is in shift mode; what to change about that?

**Comment by Be-ing** [comment]
Whenever hotcue_X_enabled changes, Mixxx needs to send MIDI messages for both the shifted and unshifted MIDI channels (status bytes 0x97 and 0x98). So try copying and pasting all the outputs for the buttons and changing 0x97 to 0x98.

The Hercules P32 controller that I previously used also has this quirky behavior. That's why I added the `sendShifted` feature in the Components library.

**Comment by nschloe** [comment]
> So try copying and pasting all the outputs for the buttons and changing 0x97 to 0x98.

That works! One needs to have
```
            <output> 
                <group>[Channel1]</group>
                <key>hotcue_1_enabled</key>
                <status>0x97</status>
                <midino>0x00</midino>
                <minimum>0.5</minimum>
            </output>
            <output>
                <group>[Channel1]</group>
                <key>hotcue_1_enabled</key>      
                <status>0x98</status>
                <midino>0x00</midino>
                <minimum>0.5</minimum>
            </output>
```
side by side.

**Comment by Be-ing** [comment]
Please add a comment to the XML file explaining why that is necessary.

**Comment by Be-ing** [comment]
I have a better idea for the effects. Instead of scrolling forward/back for focusing effects:

  * Level/Depth knob: metaknob of focused effect; if no effect is focused, controls the mix knob for effect unit 1
  * Beat `<`: focus effect 1
  * Beat `>`: focus effect 2
  * FX select button: focus effect 3
  * On/Off button: enable/disable focused effect
  * 1/2/Master switch: assign effect unit 1 to deck 1/2/master as labeled

Unfocus by pressing the button for the focused effect again (so if effect 1 is focused, press Beat `<` to unfocus and make the knob control the mix knob of the unit).

**Comment by nschloe** [comment]
Okay, I've merged more stuff from the original PR now, and more things are working now. I still need some assistance with the lights going on on shift-press though. The hotcues are fixes, but other buttons are also losing light, e.g., beat sync. How to prevent this?

**Comment by Be-ing** [comment]
I presume the same way. Send output for the shifted MIDI channel in addition to the unshifted MIDI channel.

File: res/controllers/Pioneer-DDJ-400.midi.xml [hunk 1]
```diff
@@ -506,7 +506,7 @@
             <control>
                 <description>RELOOP/EXIT (DECK1) - press - (loop off) Reloop, (loop on) Loop exit</description>
                 <group>[Channel1]</group>
-                <key>reloop_toggle</key> <!-- check if correct -->
+                <key>reloop_andstop</key> <!-- check if correct -->
```
Comment (Be-ing) [inline res/controllers/Pioneer-DDJ-400.midi.xml]: Why did you change this? I think `reloop_andstop` makes sense for the shifted press, but I don't understand why you are using it for the unshifted press too.

Comment (nschloe) [reply res/controllers/Pioneer-DDJ-400.midi.xml]: By mistake. Fixed now.

**Comment by Holzhaus** [comment]
Are you using any features new in Mixxx 2.3 in this mapping? If not, can you retarget this PR to the 2.2 branch and rebase on 2.2? This will make it possible to include this in 2.2.4 instead of 2.3.

**Comment by Holzhaus** [comment]
Are you using any features new in Mixxx 2.3 in this mapping? If not, can you retarget this PR to the 2.2 branch and rebase on 2.2? This will make it possible to include this in 2.2.4 instead of 2.3.

**Comment by Be-ing** [comment]
This PR contains a lot of unrelated commits now. It looks like you rebased onto master instead of 2.2. Do an interactive rebase `git rebase -i mixxxdj/2.2` to rebase on 2.2 and remove all commits that are not related to this mapping.

**Comment by nschloe** [comment]
>  It looks like you rebased onto master instead of 2.2.

I did indeed. I don't know how to undo either. `git rebase -i mixxxdj/2.2` just gives `fatal: invalid upstream 'mixxxdj/2.2'`. The easiest thing is probably simply to start a new PR.

**Comment by Be-ing** [comment]
Have you set up the mixxxdj repository as a remote? The [wiki](https://mixxx.org/wiki/doku.php/using_git#set_up_git) has instructions to set it up as a remote called "upstream", so the command would be `git rebase -i upstream/2.2`

**Comment by nschloe** [comment]
Yeah I always to that, but then I've never done an "interactive" rebase. I don't know what do to with the interface git gives me there.

**Comment by Be-ing** [comment]
Delete the lines for every commit that doesn't have to do with your mapping, leaving only the commits for your mapping.

**Comment by nschloe** [comment]
> Delete the lines for every commit that doesn't have to do with your mapping, leaving only the commits for your mapping.

There are hundreds of commits, and I only see the title, not the author. It's too much work. I'll simply start a new PR.

**Comment by nschloe** [comment]
See #2403.

**Comment by Be-ing** [comment]
All of your commits are at the end. It should be easy to delete the commits before that.

**Comment by nschloe** [comment]
I've played around with it for half an hour now but I can't seem to get the rebase to work. I did
```
$ git remote add upstream https://github.com/mixxxdj/mixxx.git
$ git fetch upstream
$ git rebase -i upstream/2.2
$ git push -f origin pioneer-ddj400
```
but I still have all those commits here. Gah.
